docs/jax-101/07-state.md

```diff
@@ -0,0 +1,275 @@
+
+The function to look at carefully is `update`.
+
+```{code-cell} ipython3
+:id: wQdU7DoAseW6
+
+from typing import NamedTuple
+
+class Params(NamedTuple):
+  weight: jnp.ndarray
+  bias: jnp.ndarray
+
+
+def init(rng) -> Params:
+  """Returns the initial model params."""
+  weights_key, bias_key = jax.random.split(rng)
+  weight = jax.random.normal(weights_key, ())
+  bias = jax.random.normal(bias_key, ())
+  return Params(weight, bias)
+
+
+def loss(params: Params, x: jnp.ndarray, y: jnp.ndarray) -> jnp.ndarray:
+  """Computes the least squares error of the model's predictions on x against y."""
+  pred = params.weight * x + params.bias
+  return jnp.mean((pred - y) ** 2)
+
+
+LEARNING_RATE = 0.005
+
+@jax.jit
+def update(params: Params, x: jnp.ndarray, y: jnp.ndarray) -> Params:
+  """Performs one SGD update step on params using the given data."""
+  grad = jax.grad(loss)(params, x, y)
+
+  # If we were using Adam or another stateful optimizer,
+  # we would also do something like
+  # ```
+  # updates, new_optimizer_state = optimizer(grad, optimizer_state)
+  # ```
+  # and then use `updates` instead of `grad` to actually update the params.
+  # (And we'd include `new_optimizer_state` in the output, naturally.)
+
+  new_params = jax.tree_multimap(
+      lambda param, g: param - g * LEARNING_RATE, params, grad)
+
+  return new_params
+```
+
++++ {"id": "dKySWouu2-Hu"}
+
+Notice that we manually pipe the params in and out of the update function.
+
+```{code-cell} ipython3
+:id: jQCYYy0yxO6K
+:outputId: 1f3b69d2-e90b-4065-cbcc-6422978d25c2
+
+import matplotlib.pyplot as plt
+
+rng = jax.random.PRNGKey(42)
+
+# Generate true data from y = w*x + b + noise
+true_w, true_b = 2, -1
+x_rng, noise_rng = jax.random.split(rng)
+xs = jax.random.normal(x_rng, (128, 1))
+noise = jax.random.normal(noise_rng, (128, 1)) * 0.5
+ys = xs * true_w + true_b + noise
+
+# Fit regression
+params = init(rng)
+for _ in range(1000):
+  params = update(params, xs, ys)
+
+plt.scatter(xs, ys)
+plt.plot(xs, params.weight * xs + params.bias, c='red', label='Model Prediction')
+plt.legend()
+plt.show()
+```
+
++++ {"id": "1wq3L6Xg1UHP"}
+
+## Taking it further
+
+The strategy described above is how any (jitted) JAX program must handle state. 
+
+Handling parameters manually seems fine if you're dealing with two parameters, but what if it's a neural net with dozens of layers? You might already be getting worried about two things:
+
+1) Are we supposed to initialize them all manually, essentially repeating what we already write in the forward pass definition?
+
+2) Are we supposed to pipe all these things around manually?
+
+The details can be tricky to handle, but there are examples of libraries that take care of this for you. One example
+is [Haiku](https://github.com/deepmind/dm-haiku), which allows you to write OOP-style neural nets without explicitly tracking their parameters, and then to transform them into functional programs.
```

Comment:
Instead of calling out Haiku here, how about we link to https://github.com/google/jax#neural-network-libraries instead.

Reply:
Great idea - done!